### DISCUSSION TITLE
api requests on collection

### DISCUSSION THREAD

**Discussion by realJogicodes** [pocketbase/pocketbase#4133] [Q&A]
Hi,

I tried this api policy: `sender = @request.auth.id || recipient = @request.auth.id`
intending to allow users to view their direct messages between each other.

I also tried 
`sender.id = @request.auth.id || recipient.id = @request.auth.id`

The idea is, to allow the user access if they are in the sender OR recipient field.

Sender is a relation field to users, recipient is also a relation field to users.

When the filter is in place, messages do not show up in my app when I query them. When I remove the filter, they show up but then everyboy sees everything, not good for pairwise messaging. My guess is iImessed something up but I just don't get it

**Comment by ganigeorgiev** [comment]
Double check that you are authenticated when sending the request.

If you use Next.js it could be also cache related (check their [Opting out docs](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching)).

If that doesn't help, please provide a code sample how you are initializing the SDK and how are you sending the request.

**Reply by realJogicodes** [reply]
Thanks or getting back to me.

- I am authenticated, I check server-side and redirect to login if not, code below

- I am using svelte kit for the front end and server code, not next.js

- So here's all of the relevant code that runs in order like this:

1. instantiate pocketbase as pb via JS SDK:

```
// src/lib/pocketbase.ts
import PocketBase from 'pocketbase';
import { writable } from 'svelte/store';

export const pb = new PocketBase('http://192.168.0.23:8090');

export const currentUser = writable(pb.authStore.model);

pb.authStore.onChange(() => {
	console.log('authStore change', pb.authStore.model);
	currentUser.set(pb.authStore.model);
});
```
2. in hooks, ensure availability of pb on server side for auth purposes

```
// src/hooks.ts

import { pb } from '$lib/pocketbase';

import type { Handle } from '@sveltejs/kit';

export const handle: Handle = async ({ event, resolve }) => {
	// make pb available in locals
	event.locals.pb = pb;
	event.locals.pb.authStore.loadFromCookie(event.request.headers.get('cookie') || '');

	// check auth status
	// make user available in locals
	try {
		if (event.locals.pb.authStore.isValid) {
			await event.locals.pb.collection('users').authRefresh();
			event.locals.user = structuredClone(event.locals.pb.authStore.model);
		}
	} catch (_) {
		event.locals.pb.authStore.clear();
		event.locals.user = undefined;
	}

	// prepare response
	const response = await resolve(event);
	response.headers.set('set-cookie', event.locals.pb.authStore.exportToCookie({ secure: false }));
	return response;
};

```

the chat route:
1st: server side code for auth check, 
login required to see route
```
// src/routes/chat/+page.server.ts

import { redirect } from '@sveltejs/kit';

export const load = (event) => {
	// redirect to login if user is undefined.
	if (!event.locals.user) {
		throw redirect(302, `/login`);
	} else {
		// pass
	}
};

```

handle rest on front end:

```
<script lang="ts">
	import { onMount, onDestroy } from 'svelte';
	import { pb } from '$lib/pocketbase';

	// get data from server
	export let data;
	const userId = data.user.id;
	const userAvatar = data.user.avatar;
	const userType = data.user.type;

	// initialize all messages and new message
	let messages: any[] = [];
	let newMessage: string;


	// permissions are off:
	// list: ""
	// view: ""
	// works
	// permissions are stringent:
	// "sender = @request.auth.id || recipient = @request.auth.id"
	// does not work
	// "sender.id = @request.auth.id || recipient.id = @request.auth.id"
	// does not work
	// (which one of the two is supposed to be it?)
	

	async function getMessages() {
		const messages = await pb
			.collection('messages')
			.getFullList(200, { sort: 'created', expand: 'sender,recipient' });
		// console.log('debug messages: ', messages);
		const results = messages.map((message) => {
			return {
				id: message.id,
				content: message.content,
				sender: message.sender
			};
		});
		console.log('getMessages resutl: ', results);
		return results;
	}

	onMount(async () => {
		// initial loading
		messages = await getMessages();
		// real time subscription for updates
		pb.collection('messages').subscribe('*', async (e) => {
			messages = await getMessages();
		});
	});

	onDestroy(() => {
		// unsubscribe to subscription
		pb.collection('messages').unsubscribe('*');
	});

	async function sendMessage() {
		const msgData = {
			content: newMessage,
			sender: userId,
			senderIs: userType
		};
		// send to api endpoint
		const response = await fetch('/api/store-message', {
			method: 'POST',
			body: JSON.stringify(msgData),
			headers: {
				'content-type': 'application/json'
			}
		});
		// TODO
		// if response is error, inform user
		// else pass
	}
</script>

<div>Markup ommitted, because not relevant to problem</div>
```

More info:

with premissions set, I get this on my front end console:

```
The connection to http://192.168.0.23:8090/api/realtime was interrupted while the page was loading.
[RealtimeService.ts:357:27](http://192.168.0.23:8090/node_modules/.pnpm/pocketbase@0.20.1/node_modules/pocketbase/src/services/RealtimeService.ts)
```

![Screenshot from 2024-01-12 14-19-50](https://github.com/pocketbase/pocketbase/assets/108186015/4407f18f-d5cf-4c8d-adf8-061455820fb9)

I logout/login into the app after making pocketbase API rule changes, to make sure auth is "fresh"

**Reply by ganigeorgiev** [reply]
It is safe to use a single global JS SDK instance in a SSR context **only** when authenticating server-side as admin.

For managing the regular users auh state you have to create a new instance on each request, otherwise different users will write over the same `pb.authStore`. For an example please refer to https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration.

**Reply by realJogicodes** [reply]
I see, TY. 
So I drop the export from lib/pocketbase and move instantiation into hooks.server.ts export const handle and pass the instance through?

I might have another clue too to that end. I noticed in the logs, that with my bad code, the front end code makes its auth requests as `data.auth	guest`, whereas the backend server side code has `data.auth	authRecord` for all it's requests. That's probably why this bounces
